src/rpc/blockchain.cpp

```diff
@@ -2230,6 +2230,70 @@ UniValue scantxoutset(const JSONRPCRequest& request)
     return result;
 }
 
+/**
+ * Serialize the UTXO set to a file for loading elsewhere.
+ *
+ * @see SnapshotMetadata
+ */
+UniValue dumptxoutset(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1)
+        throw std::runtime_error(
+            RPCHelpMan{
+                "dumptxoutset",
+                "\nWrite the serialized UTXO set to disk.\n",
+                {
+                    {"path", RPCArg::Type::STR, RPCArg::Optional::NO, /* default_val */ "", "path to the output file"},
+                },
+                RPCResults{},
+                RPCExamples{""},
+            }.ToString()
+        );
+
+    std::string path = request.params[0].get_str();
+    FILE* file{fsbridge::fopen(path, "wb")};
+    CAutoFile afile{file, SER_DISK, CLIENT_VERSION};
+    std::unique_ptr<CCoinsViewCursor> pcursor;
+
+    {
+        LOCK(::cs_main);
+        ::FlushStateToDisk();
+        pcursor = std::unique_ptr<CCoinsViewCursor>(::ChainstateActive()->CoinsDB()->Cursor());
+        assert(pcursor);
+    }
+
+    CCoinsStats stats;
+    if (!GetUTXOStats(::ChainstateActive()->CoinsDB(), stats)) {
+        throw JSONRPCError(RPC_INTERNAL_ERROR, "Unable to read UTXO set");
+    }
+
+    CBlockIndex* tip;
+    {
+        LOCK(::cs_main);
+        tip = LookupBlockIndex(stats.hashBlock);
+        assert(tip);
+    }
+    SnapshotMetadata metadata{
+        CDiskBlockIndex(tip), stats.hashSerialized, stats.coins_count, tip->nChainTx};
+
+    afile << metadata;
+
+    while (pcursor->Valid()) {
+        boost::this_thread::interruption_point();
+        COutPoint key;
+        Coin coin;
+        if (pcursor->GetKey(key) && pcursor->GetValue(coin)) {
+            afile << key;
```

Comment:
Yeah, this is an incorrect relic of this RPC being a hackneyed prototype. We'll probably have to lock cs_main for the duration of this call for now, but I think that'll be prohibitively lengthy when we start generating snapshots as part of normal operation, so down the road we may have to introduce a unique lock for the coinsdb (i.e. to prevent flushes from happening mid-snapshot).

We're fine to use different cursor objects so long as the leveldb data hasn't changed between calls (which it obviously won't if we hold cs_main).

Reply:
This makes me think that the existing `gettxoutsetinfo` RPC is broken since it isn't locking cs_main during the `GetUTXOStats()` call. I wonder if the author of that function assumed that leveldb handles locking as some kind of RAII associated with use of the iterator, but that's not the case ([DBImpl::NewIterator](https://github.com/bitcoin-core/leveldb/blob/bitcoin-fork/db/db_impl.cc#L1156-L1166), [DBIter](https://github.com/bitcoin-core/leveldb/blob/bitcoin-fork/db/db_iter.cc#L51-L64)).

Reply:
The cursor returned iterates over a snapshot of the database at the time of its creation; the db can be modified simultaneously, but won't affect what the cursor sees.

Reply:
@sipa I can't find any mention of this behavior in the leveldb docs or header files. I see it happening [here](https://github.com/bitcoin-core/leveldb/blob/bitcoin-fork/db/db_impl.cc#L1159), but do we want to rely on something so implicit?

Reply:
We should probably make it explicit through the Snapshot API of LevelDB. I thought we already did, actually - so I'm a bit confused how I knew this as I don't remember looking in the code.